use_parsing_cache: Handle SIGXFSZ (file size limit exceeded)

IagoAbal opened the pull request:
Closes PA-318

test plan:

    % semgrep-core -lang js -j 1 -e '$X == $X' -fast \
          -use_parsing_cache ./cache \
          ~/semgrep/perf/bench/juice-shop/input/juice-shop/
      #^ no longer gets killed by SIGXFSZ (Apple M1, MacOS 11)
      # and with -j 2 we no longer get a segfault



PR checklist:
- [ ] documentation is up to date
- [x] changelog is up to date

mjambon commented on semgrep-core/src/cli/Main.ml:
```diff
@@ -402,7 +402,23 @@ let cache_computation file cache_file_of_file f =
+                 logger#error
+                   "Could not write cache file for %s: file size limit exceeded"
+                   file;
+                 cleanup_cache ()));
```
I think this signal handler is unsafe because it's going to execute asynchronously, similar to a thread. If you want to write a message to stderr, use `Unix.write` (unbuffered write) instead of the buffered write that `logger#error` presumably uses. In other words, you probably don't want to do that at all. The simplest thing to do is to have the signal handler raise an exception, and have this exception be caught in the normal execution flow (main thread).

Also, registering this signal handler should probably be done once and for all during program initialization (wherever we do these things).

e.g.
```
exception File_size_limit_exceeded
let () =
  Sys.signal_handler Sys.sigxfsz
    (Sys.Signal_handle (fun _signal -> raise File_size_limit_exceeded))
```
and then catch `File_size_limit_exceeded` where you think it's relevant.

IagoAbal replied:
Great feedback, I'll fix that.

IagoAbal replied:
If we raise an exception inside the handler this can pop up anywhere and that is not very useful for the purpose of recovering from this error. It seems that if you simply ignore it, that will cause the sys call to fail, and you get a `Sys_error` instead, from which it is easy to catch and recover from. Let me know what do you think!

mjambon replied:
If that's how it works, let's try this.

aryx commented on semgrep-core/src/cli/Main.ml:
```diff
@@ -1458,6 +1463,14 @@ let options () =
 
 (*s: function [[Main_semgrep_core.main]] *)
 let main () =
+  (* SIGXFSZ (file size linit exceeded) by default will kill the process, which
```
typo, limit not linit.

aryx commented on semgrep-core/src/cli/Main.ml:
```diff
@@ -402,7 +402,12 @@ let cache_computation file cache_file_of_file f =
           let res = f () in
-          Common2.write_value (Version.version, file, res) file_cache;
+          (try Common2.write_value (Version.version, file, res) file_cache
+           with Sys_error err ->
```
maybe a comment here that talks about SIGXFSZ?
What is the size limit actually? Why do we reach it? Our ASTs are too big?

IagoAbal replied:
I don't know exactly how this happens but in some shells `ulimit -f` shows me `unlimited` whereas in others shows me `10000` (file blocks). I certainly have not set that limit myself, it's very strange. I get the error on a file whose AST dump has a size of 19,993,228 bytes (so not that large).

emjin replied:
Just to check, do you get this error always or when `ulimit -f` shows you `10000`?

IagoAbal replied:
If there is no file size limit then I don't get the error.

IagoAbal committed:
use_parsing_cache: Handle SIGXFSZ (file size limit exceeded)

Thanks to Martin for valuable feedback.

Closes PA-318

test plan:

    % semgrep-core -lang js -j 1 -e '$X == $X' -fast \
          -use_parsing_cache ./cache \
          ~/semgrep/perf/bench/juice-shop/input/juice-shop/
      #^ no longer gets killed by SIGXFSZ (Apple M1, MacOS 11)
      # and with -j 2 we no longer get a segfault

IagoAbal commented:
Initially I thought that we were creating huge files in some cases. It turns out it is not the case. But, somehow, something is setting a file size limit in some of my terminals triggering this problem. I wonder if it is any of our internal scripts, `pipenv`, or some other tool we use. Because that's all I'm running on this laptop.

Anyways, I will still merge this because the changes are minimal, and because it's still a good idea to be defensive, catch errors during writing, and clean up the cache files to avoid corruption. But if you dislike it feel free to revert.

mjambon commented:
+1 on being defensive
